doveadm plugin API: dsync hooks & exit_codes

errror opened the pull request:
Hello Dovecot developers,

I created this merge request to extend `doveadm`'s `sync` and `dsync-server` commands with hook functions. These hooks allow me to provide an extension for Dovecot: [dovecot-replication](https://people.mpi-klsb.mpg.de/~pcernko/dovecot-replication.shtml). In addition to the dsync hooks, the patch also adds an interface to register custom doveadm exit codes that can be transferred via the doveadm protocol.

I would appreciate if you could accept my merge request in on of the next releases of Dovecot. Feel free to squash all commits into one.

If you have questions or suggestions, feel free to ask me here on in your IRC channel at oftc.net.

cmouse commented:
Hi! Thank you for the PR. We can consider maybe adding just the hooks, but not the replication part.

errror commented:
@cmouse Yes, that was exactly my intention.

I'm planing to provide the replication part as a plugin, but that requires the hooks and exit-code API in this pull request. Feel free to squash the commits so that the references to the replication part disappear. The pull request only modifies the existing `doveadm-dsync` and `doveadm-protocol` sources and the doveadm Makefile (for an additional include path). Sorry for the replication references in my commit messages, I thought it would be nice for you to see the development progress.

errror commented on src/doveadm/doveadm-dsync.c:
```diff
@@ -53,66 +54,267 @@
+	return strcmp(s1, s2);
+}
+
+void dsync_hooks_alloc(struct dsync_cmd_context *ctx);
```
No, these functions are only used in `doveadm-dsync.c`. I had to declare them to fix a compiler warning about missing declaration. Should I move the declaration  to a header file anyway?

cmouse replied:
using `static void` helps. this should be used for all functions that are only used inside doveadm-dsync.c

errror commented:
Thanks very much, @cmouse !
I replaced the declarations of all functions only used locally with `static` in their definitions. That brought out another issue: `doveadm_exit_code_deinit()` was never used. I fixed that by calling it during doveadm deinit. All changes have been pushed to my merge request.

errror commented:
Hi Dovecot developers, hi @cmouse ,

may I ask, if there is anything else required from my side for the proposed patch to be merged? Or are you still evaluating my code changes?

Best regards,
Patrick

cmouse commented:
Can you please clean up your patch into just the changes that you would want us to merge, not the whole thing?
